File: exulanica/reconstruction/pointmap.py
"""A point map: what one photograph turns out to have been looking at.

**This module knows nothing about evidence and must not learn.** It does not import
``exulanica.evidence``, ``exulanica.store`` or ``exulanica.db``, and an
import-linter contract enforces that. Invariant 2 says reconstruction is never
evidence, and the strongest available form of that
rule is a producer that cannot name a citation: a module with no way to construct an address
cannot return one, however it is later refactored by somebody who never read this comment.

**Stdlib arrays rather than numpy.** This package's only third-party dependency is Pillow, which
the project already carries for decode and EXIF. A quarter of a million points is 4.5 MB and
``array`` writes it with one ``tobytes`` call, so numpy would buy nothing here and would put a
build dependency in the path of a stage that has to run on a laptop. The depth model that fills
these arrays may well use numpy; that is its business and it is behind an interface.

**What the alpha channel means is declared, not inferred.** The container carries
``colorAlpha`` as an enum, ``support`` or ``confidence``, and this producer writes ``support``:
what :mod:`exulanica.reconstruction.build` puts there is a spacing ratio, which is coverage rather
than belief. ADR-0010 D5 made the field an enum for exactly this reason. Both writers used to
declare ``confidence`` and one of them had stopped meaning it, and the renderer told them apart
by the presence of a statistics key, "which is a format flag that nobody declared as one".

**Two uint16 channels per point, not one.** ADR-0010 D3: the segment attribute was two bytes,
WebGPU requires a vertex stream whose element size is a multiple of four, and the binding was
paying a per-point CPU pass to widen it. The container now stores what the engine already padded
to. Channel 0 is the segment id, unchanged in meaning; channel 1 is a flags word whose bit 0
says this point lost a neighbour to the silhouette drop, and whose remaining bits are reserved
and validated zero.
"""

from __future__ import annotations

from array import array
from dataclasses import dataclass, field
from typing import Final

__all__ = [
    "MAX_SEGMENT_ID",
    "POINT_STRIDE_BYTES",
    "RESERVED_TAG_FLAGS",
    "TAG_ONE_SIDED",
    "PointMap",
    "Segment",
]

#: Bytes per point on the wire: position 12, colour 4, tags 4. Eighteen under OPM/1, and the two
#: extra bytes are the alignment the WebGPU binding was already paying a per-point CPU pass for.
POINT_STRIDE_BYTES: Final = 20

#: Bit 0 of the tags flags channel: a four-neighbour of this point was dropped by the silhouette
#: test, so the sample it would have formed a tangent frame against is not in the file.
#:
#: The renderer can already see that a grid cell beside this one is empty. What it cannot see is
#: WHY, and the two answers want opposite treatments: a cell the model never placed is where the
#: observed surface honestly ends, and a cell the silhouette test removed is a surface that
#: continues with its rim taken off. ADR-0010 D4 carries the distinction as one bit rather than
#: as a normal or a radius, on the measurement that load-time tangent frames beat both.
TAG_ONE_SIDED: Final = 0x0001

#: Every bit of the flags channel that is not yet defined. Reserved and validated zero by
#: :func:`exulanica.reconstruction.validation.validate_opm`, so the next flag is a widening of a
#: declared word rather than a reinterpretation of bytes some writer already filled with
#: something else.
RESERVED_TAG_FLAGS: Final = 0xFFFE

#: The largest segment id the container declares, which is the largest the renderer can draw.
#:
#: ADR-0010 D3: "The declared range of a segment id is bounded by what the renderer can actually
#: draw, not by the width of the field." The semantic table in
#: ``web/packages/atlas-react/src/playcanvas/semantics.ts`` holds sixteen entries, the binding
#: raises for a larger id and the shaders index a fixed array, so a file declaring id 900 would
#: be declaring something no renderer in this product can present. The field stays sixteen bits
#: wide because the binding shape is uint16x2; what is bounded is the declared range, and
#: widening it is a change to the table and this constant together.
MAX_SEGMENT_ID: Final = 15

#: Segment 0 is what a producer that does not segment anything writes. It is named `unsegmented`
#: rather than `unknown`, because nothing looked and failed: nothing looked.
UNSEGMENTED: Final = 0


@dataclass(frozen=True, slots=True)
class Segment:
    """A semantic region of the point map, if the producer has one. Often just the default."""

    id: int
    name: str
    cls: str


@dataclass(slots=True)
class PointMap:
    """Points in the capture's own local frame, metres, +Y up, -Z forward.

    The frame is the CAPTURE's, not the Atlas's. A point map has a real metric scale and a real
    origin at the camera; where that sits in the Atlas is a presentation decision made elsewhere
    and carried by the island's placement. Conflating the two is risk R-48 and the reason
    `atlas-core` does not export a distance function over atlas positions.
    """

    #: 3 floats per point.
    position: array
    #: 4 bytes per point, rgba, where a is whatever ``colorAlpha`` declares. This producer
    #: writes support.
    color: bytearray
    #: 2 uint16 per point, interleaved: segment id then flags. Interleaved rather than held as
    #: two arrays because it is one vertex attribute of two components, and splitting it here
    #: would mean a per-point pass to weave it back together on the way into the file.
    tags: array
    segments: list[Segment] = field(default_factory=list)
    #: What the producer measured about itself. Never a plausible number.
    statistics: dict[str, float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if self.position.typecode != "f":
            raise ValueError("position must be an array('f')")
        if self.tags.typecode != "H":
            raise ValueError("tags must be an array('H')")
        count = len(self.position) // 3
        if len(self.position) != count * 3:
            raise ValueError("position is not a whole number of points")
        if len(self.color) != count * 4:
            raise ValueError(f"colour has {len(self.color)} bytes for {count} points")
        if len(self.tags) != count * 2:
            raise ValueError(f"tags has {len(self.tags)} entries for {count} points")
        # Shape only, and no per-point pass. What is IN the two channels is checked by
        # `validate_opm` over the encoded bytes, which is the check that stands between a
        # producer and a durable artifact and which already walks every point once. A second
        # walk here would double the cost of building a map to re-answer a question at the
        # weaker boundary: this constructor is reachable only from inside this package.

    @property
    def count(self) -> int:
        return len(self.position) // 3

    def bounds(self) -> tuple[tuple[float, float, float], tuple[float, float, float]]:
        """The axis-aligned extent, or a degenerate box at the origin for an empty map.

        A degenerate box rather than an exception: an empty point map is a real outcome for a
        photograph the model found nothing measurable in, and the caller that has to decide what
        to do about that is the quality gate, not this.
        """
        if self.count == 0:
            return ((0.0, 0.0, 0.0), (0.0, 0.0, 0.0))
        lo = [float("inf")] * 3
        hi = [float("-inf")] * 3
        for index in range(0, len(self.position), 3):
            for axis in range(3):
                value = self.position[index + axis]
                if value < lo[axis]:
                    lo[axis] = value
                if value > hi[axis]:
                    hi[axis] = value
        return ((lo[0], lo[1], lo[2]), (hi[0], hi[1], hi[2]))

    def packed_bytes(self) -> bytes:
        """The three sections back to back, in the order a renderer reads them.

        Position, then colour, then tags, with no padding between them. That contiguity is
        what lets the PlayCanvas binding hand the engine a zero-copy view of the file: it holds
        one vertex buffer per mesh and computes its own offsets for a tightly packed planar
        format, so a gap anywhere in here costs a per-point CPU repack that the binding correctly
        reports rather than hides.

        The ORDER is the container's, and it is asserted in the header rather than assumed here:
        :func:`exulanica.reconstruction.opm.encode_opm` reads the same section table this follows,
        so a reordering that reached one of the two would fail its own validator.
        """
        return self.position.tobytes() + bytes(self.color) + self.tags.tobytes()

File: web/packages/scene-synth/src/pointmap.ts
import type { Intrinsics } from './camera.js';
import { pixelRay } from './camera.js';
import { fbm3 } from './noise.js';
import type { Primitive, Segment } from './primitives.js';
import { hash2Unit } from './rng.js';
import type { DepthBuffers } from './raster.js';

/**
 * Unproject the surviving pixels into a point map.
 *
 * PER-POINT COLOUR and PER-POINT SEGMENT ID are both required by the brief so that per-point
 * semantic behaviour (dissolve, highlight) can be driven by real labels later rather than by a
 * shader guessing. The segment id is the primitive's semantic class, so it is a real label and
 * not a colour bucket.
 *
 * CONFIDENCE RIDES IN THE COLOUR ALPHA. That is a deliberate choice and not a saving of two
 * bytes: both candidate engines support a normalised RGBA vertex colour attribute out of the
 * box, so an unmodified renderer draws a low-confidence point fainter with no shader work at
 * all. "An unconfirmed candidate must LOOK unconfirmed" then holds by default rather than by
 * remembering to implement it. The header records `colorAlpha: "confidence"` so nobody has to
 * infer it, and under OPM/2 that field is an enum whose other value is `support`: this generator
 * writes a belief and the reconstruction path writes coverage, and a reader is now told which.
 *
 * THE SEGMENT ID SHARES AN ATTRIBUTE WITH A FLAGS WORD. ADR-0010 D3 widened the container's
 * two-byte segment channel to a four-byte `tags` section of two uint16 channels, because WebGPU
 * rejects a vertex stream whose arrayStride is not a multiple of 4 and the renderer binding was
 * paying a per-point CPU pass to widen it. Channel 1 carries bit 0 of D4, whether this point had
 * a neighbour carved at an occlusion boundary.
 */

/**
 * Bit 0 of a point's flags channel: a four-neighbour was carved at an occlusion boundary.
 *
 * Declared beside the type that carries the channel rather than in `format/opm.ts`, which is the
 * same placement `exulanica/reconstruction/pointmap.py` has and the arrangement that keeps the
 * dependency one-way: the writer reads the point map, and the point map knows nothing about the
 * container it will be written into.
 */
export const TAG_ONE_SIDED = 0x0001;

export interface PointMap {
  readonly count: number;
  /** 3 floats per point, LOCAL frame, metres, +Y up, -Z forward. */
  readonly position: Float32Array;
  /** 4 bytes per point: R, G, B, confidence. */
  readonly color: Uint8Array;
  /** 2 uint16 per point, interleaved: segment id then flags. */
  readonly tags: Uint16Array;
  readonly min: readonly [number, number, number];
  readonly max: readonly [number, number, number];
}

const SUN: readonly [number, number, number] = [0.4204, 0.7207, 0.5506];
const AMBIENT = 0.34;
const HAZE: readonly [number, number, number] = [0.55, 0.615, 0.70];
const HAZE_DENSITY = 0.0072;
const GAMMA = 1 / 2.2;

const toByte = (v: number): number =>
  Math.max(0, Math.min(255, Math.round(Math.pow(Math.max(0, Math.min(1, v)), GAMMA) * 255)));

export function buildPointMap(
  buf: DepthBuffers,
  k: Intrinsics,
  keep: Uint8Array,
  confidence: Uint8Array,
  oneSided: Uint8Array,
  count: number,
  segments: readonly Segment[],
  prims: readonly Primitive[],
  eye: readonly [number, number, number],
  seed: number,
  depthJitter: number,
): PointMap {
  const position = new Float32Array(count * 3);
  const color = new Uint8Array(count * 4);
  const tags = new Uint16Array(count * 2);
  const [ox, oy, oz] = eye;

  let min: [number, number, number] = [Infinity, Infinity, Infinity];
  let max: [number, number, number] = [-Infinity, -Infinity, -Infinity];
  let w = 0;

  for (let py = 0; py < buf.height; py += 1) {
    for (let px = 0; px < buf.width; px += 1) {
      const i = py * buf.width + px;
      if (keep[i] === 0) continue;

      // High-frequency per-pixel depth jitter, applied last. Low-frequency bias was applied in
      // the raster pass, before edge detection, so it could not manufacture false silhouettes.
      const jitter = 1 + (hash2Unit(px, py, seed ^ 0x1c9d) - 0.5) * 2 * depthJitter;
      const d = buf.depth[i]! * jitter;

      const { dx, dy } = pixelRay(k, px, py);
      const hx = ox + dx * d;
      const hy = oy + dy * d;
      const hz = oz - d;

      position[w * 3] = hx;
      position[w * 3 + 1] = hy;
      position[w * 3 + 2] = hz;
      if (hx < min[0]) min[0] = hx;
      if (hy < min[1]) min[1] = hy;
      if (hz < min[2]) min[2] = hz;
      if (hx > max[0]) max[0] = hx;
      if (hy > max[1]) max[1] = hy;
      if (hz > max[2]) max[2] = hz;

      const seg = segments[prims[buf.prim[i]!]!.segment]!;
      tags[w * 2] = seg.id;
      tags[w * 2 + 1] = oneSided[i] === 1 ? TAG_ONE_SIDED : 0;

      const nx = buf.normal[i * 3]! / 127;
      const ny = buf.normal[i * 3 + 1]! / 127;
      const nz = buf.normal[i * 3 + 2]! / 127;
      const lambert = Math.max(0, nx * SUN[0] + ny * SUN[1] + nz * SUN[2]);
      const light = AMBIENT + (1 - AMBIENT) * lambert;

      // Surface texture, in world space so it does not swim between ladder rungs.
      const detail = fbm3(hx * 1.7, hy * 1.7, hz * 1.7, seed ^ 0x3ef1, 4);
      const mod = 1 + (detail - 0.5) * 0.95 * seg.texture;

      const fog = 1 - Math.exp(-d * HAZE_DENSITY);
      for (let c = 0; c < 3; c += 1) {
        const lit = seg.albedo[c]! * light * mod;
        color[w * 4 + c] = toByte(lit * (1 - fog) + HAZE[c]! * fog);
      }
      color[w * 4 + 3] = confidence[i]!;

      w += 1;
      if (w === count) break;
    }
    if (w === count) break;
  }

  if (w !== count) {
    throw new Error(`point map wrote ${w} points but the mask said ${count}`);
  }

  return { count, position, color, tags, min, max };
}
